mmdet3d/core/bbox/coders/km3d_decoder.py

```diff
@@ -0,0 +1,451 @@
+    """
+
+    # expand ind with the third dimension to be the same as feat
+    dim  = feat.size(2)
+    ind  = ind.unsqueeze(2).expand(ind.size(0), ind.size(1), dim)
+    feat = feat.gather(1, ind)
+    if mask is not None:
+        mask = mask.unsqueeze(2).expand_as(feat)
+        feat = feat[mask]
+        feat = feat.view(-1, dim)
+    return feat
+
+
+def _transpose_and_gather_feat(feat, ind):
+    """gather feat with assigned index, with feat be permuted with (0,2,3,1) and reshaped to three dimensions.
+
+        Args:
+            heat (torch.Tensor): Predictions from model, batchsize * channels * h * w
+            ind (torch.Tensor): index to be gathered
+
+        Returns:
+            torch.Tensor: feat, batchsize * channels * h * w
+    """
+    feat = feat.permute(0, 2, 3, 1).contiguous()
+    feat = feat.view(feat.size(0), -1, feat.size(3))
+    feat = _gather_feat(feat, ind)
+    return feat
+
+
+
+def _topk(scores, K=40):
+    """Select top k batches of scores, index, classes, y's, and x's given scores.
+
+        Args:
+            scores (torch.Tensor): batchsize * channels * h * w
+            K (int): top K value
+
+        Returns:
+            torch.Tensor: batchsize * K, top k scores
+            torch.Tensor: batchsize * K, top k index
+            torch.Tensor: batchsize * K, top k classes
+            torch.Tensor: batchsize * K, top k y's
+            torch.Tensor: batchsize * K, top k x's
+    """
+    batch, cat, height, width = scores.size()
+
+    topk_scores, topk_inds = torch.topk(scores.view(batch, cat, -1), K)
+
+    topk_inds = topk_inds % (height * width)
+    topk_ys = (topk_inds / width).int().float()
+    topk_xs = (topk_inds % width).int().float()
+
+    topk_score, topk_ind = torch.topk(topk_scores.view(batch, -1), K)
+    topk_clses = (topk_ind / K).int()
+    topk_inds = _gather_feat(
+        topk_inds.view(batch, -1, 1), topk_ind).view(batch, K)
+    topk_ys = _gather_feat(topk_ys.view(batch, -1, 1), topk_ind).view(batch, K)
+    topk_xs = _gather_feat(topk_xs.view(batch, -1, 1), topk_ind).view(batch, K)
+
+    return topk_score, topk_inds, topk_clses, topk_ys, topk_xs
+
+
+def gen_position(kps, dim, rot, meta, const):
+    """generate position from 2D frame to world frame.
+
+        Args:
+            kps (torch.Tensor): 1 * batchsize * 18
+            dim (int): 1 * batchsize * 3 (h, w, l)
+            rot (torch.Tensor): 1* batchsize * 8
+            meta (dict): should contain keys below.
+                - trans_output_inv: transition parameters for computing A in pinv's matrix.
+                - calib: calibration for computing A in pinv's matrix.
+            const (torch.Tensor): 1 * batchsize * 16 * 2
+
+        Returns:
+            torch.Tensor: 1 * batchsize * 3: pinv, coordinates in world frame
+            torch.Tensor: 1 * batchsize * 1: rot_y, yaw angle
+            torch.Tensor: 1 * batchsize * 18: kps, key points coordinates
+    """
+    b = kps.size(0)  # 1
```

Comment:
we need to avoid single alphabet variables, use meaningful names like batch_size, channels.